src/devices/vdml_gps.c

```diff
@@ -110,34 +110,70 @@ gps_position_s_t gps_get_position(uint8_t port) {
 	return_port(port - 1, rtv);
 }
 
-double gps_get_heading(uint8_t port) {
-	claim_port_f(port - 1, E_DEVICE_GPS);
-	double rtv = vexDeviceGpsDegreesGet(device->device_info);
+double gps_get_position_x(uint8_t port) {
+    claim_port_f(port - 1, E_DEVICE_GPS);
+	V5_DeviceGpsAttitude data;
+	vexDeviceGpsAttitudeGet(device->device_info, &data, false);
+	double rtv = data.position_x;
 	return_port(port - 1, rtv);
 }
 
-double gps_get_heading_raw(uint8_t port) {
-	claim_port_f(port - 1, E_DEVICE_GPS);
-	double rtv = vexDeviceGpsHeadingGet(device->device_info);
+double gps_get_position_y(uint8_t port) {
+    claim_port_f(port - 1, E_DEVICE_GPS);
+	V5_DeviceGpsAttitude data;
+	vexDeviceGpsAttitudeGet(device->device_info, &data, false);
+	double rtv = data.position_y;
 	return_port(port - 1, rtv);
 }
 
-double gps_get_rotation(uint8_t port) {
-	claim_port_f(port - 1, E_DEVICE_GPS);
-	double rtv = vexDeviceGpsRotationGet(device->device_info);
+gps_orientation_s_t gps_get_orientation(uint8_t port) {
+    gps_orientation_s_t rtv = {PROS_ERR_F, PROS_ERR_F, PROS_ERR_F};
+	if (!claim_port_try(port - 1, E_DEVICE_GPS)) {
+		return rtv;
+	}
+	v5_smart_device_s_t* device = registry_get_device(port - 1);
+	V5_DeviceGpsAttitude data;
+	vexDeviceGpsAttitudeGet(device->device_info, &data, false);
+	rtv.pitch = data.pitch;
+	rtv.roll = data.roll;
+	rtv.yaw = data.yaw;
 	return_port(port - 1, rtv);
 }
 
-int32_t gps_set_rotation(uint8_t port, double target) {
-	claim_port_i(port - 1, E_DEVICE_GPS);
-	vexDeviceGpsRotationSet(device->device_info, target);
-	return_port(port - 1, PROS_SUCCESS);
+double gps_get_pitch(uint8_t port) {
+    claim_port_f(port - 1, E_DEVICE_GPS);
+	V5_DeviceGpsAttitude data;
+	vexDeviceGpsAttitudeGet(device->device_info, &data, false);
+	double rtv = data.pitch;
+	return_port(port - 1, rtv);
 }
 
-int32_t gps_tare_rotation(uint8_t port) {
-	claim_port_i(port - 1, E_DEVICE_GPS);
-	vexDeviceGpsRotationSet(device->device_info, 0);
-	return_port(port - 1, PROS_SUCCESS);
+double gps_get_roll(uint8_t port) {
+    claim_port_f(port - 1, E_DEVICE_GPS);
+	V5_DeviceGpsAttitude data;
+	vexDeviceGpsAttitudeGet(device->device_info, &data, false);
+	double rtv = data.roll;
+	return_port(port - 1, rtv);
+}
+
+double gps_get_yaw(uint8_t port) {
+    claim_port_f(port - 1, E_DEVICE_GPS);
+	V5_DeviceGpsAttitude data;
+	vexDeviceGpsAttitudeGet(device->device_info, &data, false);
+	double rtv = data.yaw;
+	return_port(port - 1, rtv);
+}
+
+double gps_get_heading(uint8_t port) {
+	claim_port_f(port - 1, E_DEVICE_GPS);
+	double rtv = vexDeviceGpsDegreesGet(device->device_info);
+	return_port(port - 1, rtv);
+}
+
+double gps_get_heading_raw(uint8_t port) {
+	claim_port_f(port - 1, E_DEVICE_GPS);
+	double rtv = vexDeviceGpsHeadingGet(device->device_info);
+	return_port(port - 1, rtv);
 }
 
 gps_gyro_s_t gps_get_gyro_rate(uint8_t port) {
```

Comment:
I know this is completely my fault but can we see if typecasting V5_DeviceGPSRaw to the return type works, I know what I did here was super scuffed and I'd appreciate if we just casted it instead of doing the extraneous copy.

Reply:
Can I have more context on this? vexDeviceGpsHeadingGet() already returns a double value, I don't see a need for typecasting.

Reply:
Can I have more context on this? vexDeviceGpsHeadingGet() already returns a double value, I don't see a need for typecasting.